Restrict UseStaticImport to static methods (#3661)

Bananeweizen opened the pull request:
* add tests for good and bad cases
* fix issue by checking method to be static
* verified that the bad case fails without the fix and passes with the fix

### Checklist
- [X] I've added unit tests to cover both positive and negative cases
- [X] I've added the license header to any new files through `./gradlew licenseFormat`
- [ ] I've used the IntelliJ IDEA auto-formatter on affected files

knutwannheden commented on rewrite-java/src/main/java/org/openrewrite/java/UseStaticImport.java:
```diff
@@ -58,7 +59,7 @@ private class UseStaticImportVisitor extends JavaIsoVisitor<ExecutionContext> {
             MethodMatcher methodMatcher = new MethodMatcher(methodPattern);
             J.MethodInvocation m = super.visitMethodInvocation(method, ctx);
-            if (methodMatcher.matches(m)) {
+            if (methodMatcher.matches(m) && method.getMethodType().hasFlags(Flag.Static)) {
```
Would be better to place this in the next condition, where we know the method type to be non-null. Also, it should be retrieved from `m` rather than `method`, so that it also supports nested calls.

timtebeek replied:
There's an unusual test which fails when we add a limitation to static methods in this recipe https://github.com/openrewrite/rewrite/blob/e5c77d5afe17070474d7b5cf6e728600c5f776af/rewrite-java-test/src/test/java/org/openrewrite/java/UseStaticImportTest.java#L76-L78

timtebeek commented on rewrite-java/src/main/java/org/openrewrite/java/UseStaticImport.java:
```diff
@@ -58,7 +59,7 @@ private class UseStaticImportVisitor extends JavaIsoVisitor<ExecutionContext> {
             MethodMatcher methodMatcher = new MethodMatcher(methodPattern);
             J.MethodInvocation m = super.visitMethodInvocation(method, ctx);
-            if (methodMatcher.matches(m)) {
+            if (methodMatcher.matches(m) && method.getMethodType().hasFlags(Flag.Static)) {
                 if (m.getMethodType() != null) {
```
```suggestion
            if (methodMatcher.matches(m)) {
                if (m.getMethodType() != null && m.getMethodType().hasFlags(Flag.Static)) {
```

timtebeek replied:
Apply suggestions from https://github.com/openrewrite/rewrite/pull/3662/files#r1382445827

timtebeek replied:
Turns out it's a little more complicated when we look at 
https://github.com/openrewrite/rewrite/blob/e5c77d5afe17070474d7b5cf6e728600c5f776af/rewrite-java-test/src/test/java/org/openrewrite/java/UseStaticImportTest.java#L76-L78

timtebeek commented on rewrite-java/src/test/java/org/openrewrite/java/UseStaticImportTest.java:
I'm going to move these tests to https://github.com/openrewrite/rewrite/blob/e5c77d5afe17070474d7b5cf6e728600c5f776af/rewrite-java-test/src/test/java/org/openrewrite/java/UseStaticImportTest.java#L30

Bananeweizen replied:
Sure. I must admit I'm not really certain how things are structured in the project. I didn't discover the tests in that subproject before. You may want to incorporate the suggestion of Knut when moving.

timtebeek commented:
Capturing the failure here to save folks a click
```
UseStaticImportTest > methodInvocationsHavingNullSelect() FAILED
    org.opentest4j.AssertionFailedError: [Unexpected result in "test/Test.java":
    diff --git a/test/Test.java b/test/Test.java
    index 462a2c3..259aa9b 100644
    --- a/test/Test.java
    +++ b/test/Test.java
    @@ -1,8 +1,5 @@ 
     package test;
 
    -import static asserts.Assert.assertEquals;
    -import static asserts.Assert.assertTrue;
    -
     class Test {
         void test() {
             assertTrue(true);
    ] 
    expected: 
      "package test;
  
      import static asserts.Assert.assertEquals;
      import static asserts.Assert.assertTrue;
  
      class Test {
          void test() {
              assertTrue(true);
              assertEquals(1, 2);
          }
      }"
     but was: 
      "package test;
  
      class Test {
          void test() {
              assertTrue(true);
              assertEquals(1, 2);
          }
      }"
```

Bananeweizen commented:
I rebased, squashed, and investigated the failing test `methodInvocationsHavingNullSelect()`. To me that looks like a bug in some other area of the code. `Assert.assert(boolean)` is marked as static initially (flags=9). But when reaching the breakpoint in this recipe, the method is no longer marked static (flags=1).  I believe that https://github.com/openrewrite/rewrite/blob/eccdbd4e229b9fc057dd5caaa8b7d9f088a4462d/rewrite-java-test/src/test/java/org/openrewrite/java/UseStaticImportTest.java#L167 is the culprit. It somehow recreates the method, but seems to miss copying the flag. Since I have no clue of how to work with the generated (private) with-method from lombok in IntelliJ, I'm not sure how to fix that.

timtebeek commented:
I'm looking at that test now and I'm inclined to say the test is quite odd; it eagerly replaces calls to instance methods with calls to static methods called within those instance methods, based on the `.withExtends(null)` and `.withDeclaringType(JavaType.ShallowClass.build("asserts.Assert"))`. That's not what the `UseStaticImport` recipe should be about, and the check on methods being static here correctly makes that fail in my opinion. Would you agree there @knutwannheden ?

https://github.com/openrewrite/rewrite/blob/eccdbd4e229b9fc057dd5caaa8b7d9f088a4462d/rewrite-java-test/src/test/java/org/openrewrite/java/UseStaticImportTest.java#L150-L220

knutwannheden commented:
That test indeed looks rather peculiar. It also looks like it is very old.

timtebeek commented:
Then let's remove it; if it was necessary even after this change I suspect something downstream will pick that up.

timtebeek commented:
Thanks a lot yet again @Bananeweizen !
